tests/test_interpreter.py

```diff
@@ -89,6 +94,82 @@ def test_binary_name_matching(self):
         for name in valid_binary_names:
             assert matches(name), "Expected {} to be valid binary name".format(name)
+
+    def test_iter_interpreter_some(self, test_interpreter1, test_interpreter2):
+        # type: (str, str) -> None
+        assert [
+            PythonInterpreter.from_binary(test_interpreter1),
+            PythonInterpreter.from_binary(test_interpreter2),
+        ] == list(PythonInterpreter.iter_candidates(paths=[test_interpreter1, test_interpreter2]))
+
+    def test_iter_interpreter_none(self):
+        # type: () -> None
+        assert [] == list(PythonInterpreter.iter_candidates(paths=[os.devnull]))
+
+    @pytest.fixture
+    def invalid_interpreter(self):
+        with temporary_dir() as bin_dir:
+            invalid_interpreter = os.path.join(bin_dir, "python")
+            touch(invalid_interpreter)
+            yield invalid_interpreter
+
+    def assert_error(self, result, expected_python):
+        # type: (InterpreterOrError, str) -> None
+        assert isinstance(result, tuple)
+        python, error_message = result
+        assert expected_python == python
+        assert isinstance(error_message, str)
+        assert len(error_message) > 0
+
+    def test_iter_interpreter_errors(self, invalid_interpreter):
+        # type: (str) -> None
+        results = list(PythonInterpreter.iter_candidates(paths=[invalid_interpreter]))
+
+        assert len(results) == 1
+        self.assert_error(results[0], invalid_interpreter)
+
+    def test_iter_interpreter_mixed(
+        self, test_interpreter1, test_interpreter2, invalid_interpreter
+    ):
+        # type: (str, str, str) -> None
+        results = list(
+            PythonInterpreter.iter_candidates(
+                paths=[test_interpreter1, invalid_interpreter, test_interpreter2]
+            )
+        )
+
+        assert len(results) == 3
+        assert [
+            PythonInterpreter.from_binary(path) for path in (test_interpreter1, test_interpreter2)
+        ] == [result for result in results if isinstance(result, PythonInterpreter)]
+        errors = [result for result in results if not isinstance(result, PythonInterpreter)]
+        assert len(errors) == 1
+        self.assert_error(errors[0], invalid_interpreter)
+
+    def test_iter_interpreter_path_filter(self, test_interpreter1, test_interpreter2):
+        assert [PythonInterpreter.from_binary(test_interpreter2)] == list(
+            PythonInterpreter.iter_candidates(
+                paths=[test_interpreter1, test_interpreter2],
+                path_filter=lambda path: path == test_interpreter2,
+            )
+        )
+
+    def test_iter_interpreter_path_filter_symlink(self, test_interpreter1, test_interpreter2):
+        with temporary_dir() as bin_dir:
+            os.symlink(test_interpreter2, os.path.join(bin_dir, "jake"))
+
+            # Verify path filtering happens before interpreter resolution, which os.path.realpaths
```

Comment:
This is the use-case coming in a follow-up PR to close #1043. The PEX runtime needs to be able to filter by basename when `PEX_PYTHON=python2` (for example) is in the environment. Today this is handled with dedicated code in `pex_bootstrapper.py` that is a near duplicate of the relevant code in `PythonInterpreter.iter_candidates`.